### PR TITLE
Change training parameters

### PR DESCRIPTION
**Pull request by mdrs-thiago**
Hi there!

I proposed this pull request to make the changing parameters related exclusively to the training procedure easier. Some neural network lectures that present the Self-Organizing Maps very often describe the training step as two-phase procedures: the ordering and the fine-tuning phases. The difference between those phases is related to training parameters, such as learning rate and the neighborhood radius (and the decay function). 

My first pull request only adds two parameters (learning_rate and decay_function) to the training functions, without removing the variables statement in __init__. I did this due to conflicts in pytest, since it modifies the object creation substantially. This modification just turns easier the modification of the parameters, given that previously I had to change in a non-intuitive way (e.g. change learning rate using som._learning_rate = new_lr). 

If you agree, I think that we could split the variable initialization into two categories. The first one, related to the training procedure (such as learning rate, sigma, neighborhood function, and decay function) could be declared while calling a train function. All other parameters (mainly related to the model architecture) could be defined as it is. This may help to not only replicate the two-phase training as cited above but also make it easier to continue the training from a previous checkpoint using different parameters. However, this change is not so simple (it may change the examples and the unit test as well), so I think that this first pull request may serve as the introduction of a proposal of alteration. 

Kind regards,
Thiago Medeiros.

### CODE REVIEW

**Comment by JustGlowing** [comment]
Hi Thiago,

Thanks for your contribution. Unfortunately the changes that you suggest are not implementable for the following reasons:

- They will break backward compatibility.
- There are better ways to improve the training from a checkpoint (e.g. save the parameters at that specific stage of the training).
- I'm not sure what you mean by "two-phase training" but fine tuning is pretty easy with the current interface.

Regards
